src/prefect_server/api/artifacts.py

```diff
@@ -0,0 +1,80 @@
+from prefect import models
+from prefect.utilities.plugins import register_api
+
+
+@register_api("artifacts.create_task_run_artifact")
+async def create_task_run_artifact(
+    task_run_id: str,
+    kind: str,
+    data: dict,
+    tenant_id: str = None,
+) -> str:
+    """
+    Create a task run artifact.
+
+    Args:
+        - task_run_id (str): the task run id
+        - kind (str): the artifact kind
+        - data (dict): the artifact data
+        - tenant_id (str, optional): the tenant id that this artifact belongs to. Defaults
+            to the tenant ID linked to the task run
+
+    Returns:
+        - str: the task run artifact ID
+    """
+    task_run = await models.TaskRun.where(id=task_run_id).first({"tenant_id"})
+
+    if not task_run:
+        raise ValueError(f"Creating artifact failed for task run ID {task_run_id}")
+
+    if not tenant_id:
+        tenant_id = task_run.tenant_id
+
+    # Insert task run artifact
+    return await models.TaskRunArtifact(
+        tenant_id=tenant_id, task_run_id=task_run_id, kind=kind, data=data
+    ).insert()
+
+
+@register_api("artifacts.update_task_run_artifact")
+async def update_task_run_artifact(
+    task_run_artifact_id: str,
+    data: dict,
+) -> bool:
+    """
+    Update a task run artifact.
+
+    Args:
+        - task_run_artifact_id (str): the task run artifact id
+        - data (dict): the artifact data
+
+    Returns:
+        - bool: if the update was successful
+    """
+    if not task_run_artifact_id:
+        raise ValueError("Must supply a valid task run artifac ID to update")
+
+    result = await models.TaskRunArtifact.where(id=task_run_artifact_id).update(
+        set={"data": data}
```

Comment:
I don't necessarily have pushback, but curious why you chose a full update API instead of an append API?

Reply:
I'm curious what you mean by an append API

Reply:
Right now if a task run wants to publish two artifacts, it'll need to collect them _both_ client-side and call `update_task_run_artifact` with the full payload.  

Alternatively, we _could_ change the API behavior so that calling `update_task_run_artifact` allows users to add multiple artifacts in sequence (this might require an implicit schema-change to the JSONB in the `data` column too).

Reply:
Perhaps I'm misunderstanding what was done but the mapping for task_run -> artifacts is one -> many so they actually call `create_task_run_artifact` each time they want to add/append an artifact and `update_task_run-artifact` exists solely so an already created task run artifact can have its data replaced (for which an example use-case is a progress bar).

Reply:
I don't think we should support append behavior for the update. If a task is going to publish two artifacts then it is going to have to call `create` twice anyways in order to create the two artifacts. Client side enforcement of how the data should be updated is probably better than reading the data and determining what to append. 

A downside of append behavior I could think of is something like this: if I have a piece of data that is of `markdown` type and I want to update where do I append the data? Is it a new entry in the data blob or do I merge the two strings together? Now pieces like the UI have to decide how to render artifacts with multiple markdown values. (which one gets displayed first, etc.) I could see the same thing with link artifacts where ideally it should be one link artifact = one link, overwrite only and we don't have to have any sense of resolution when it comes to merging the data.

Reply:
Ahhh yes thank you both for clarifying; I am 100% on the same page now and definitely agree your current API is the way to go 👍